src/modm/platform/pwm/sam_x7x/pwm_base.hpp.in

```diff
@@ -0,0 +1,298 @@
+	enum class TcTriggerSource
+	{
+		Comparator = 0,
+		CounterEvent = PWM_CMR_TCTS,
+
+		Mask = CounterEvent
+	};
+
+	enum class DeadTimeGeneration
+	{
+		Disabled = 0,
+		Enabled = PWM_CMR_DTE,
+
+		Mask = Enabled
+	};
+
+	/// Invert output polarity (after dead-time generator, no effect on override mode)
+	enum class OutputPolarity
+	{
+		PwmHInverted = PWM_CMR_DTHI,
+		PwmLInverted = PWM_CMR_DTLI,
+
+		Mask = (PWM_CMR_DTHI | PWM_CMR_DTLI)
+	};
+
+	enum class PushPullMode
+	{
+		Disabled = 0,
+		Enabled = PWM_CMR_PPM,
+
+		Mask = Enabled
+	};
+
+	/**
+	 * Channel mode config
+	 *
+	 * Example usage:
+	 * \code
+	 * constexpr auto mode = Pwm{{id}}::ChannelMode()
+	 *   .setClock(Pwm{{id}}::ChannelClock::Mck)
+	 *   .setDeadTimeGeneration(Pwm{{id}}::DeadTimeGeneration::Enabled);
+	 *
+	 * Pwm{{id}}::setChannelMode(Pwm{{id}}::Channel::Ch1, mode);
+	 * \endcode
+	 */
+	struct ChannelMode
+	{
+		/// Set channel clock source
+		constexpr ChannelMode
+		setClock(ChannelClock clock);
+
+		/// Read channel clock source
+		constexpr ChannelClock
+		clock() const;
+
+		/// Set channel alignment (left / center-aligned)
+		constexpr ChannelMode
+		setAlignment(ChannelAlignment alignment);
+
+		/// Read channel alignment (left / center-aligned)
+		constexpr ChannelAlignment
+		alignment() const;
+
+		/// Set channel comparator polarity
+		constexpr ChannelMode
+		setComparatorPolarity(ComparatorPolarity polarity);
+
+		/// Read channel comparator polarity
+		constexpr ComparatorPolarity
+		comparatorPolarity() const;
+
+		/// Configure center-aligned mode counter event and update at half-cycle
+		/// Unused in left-aligned mode.
+		constexpr ChannelMode
+		setHalfPeriodMode(HalfPeriodMode mode);
+
+		/// Read center-aligned mode counter event and update at half-cycle
+		constexpr HalfPeriodMode
+		halfPeriodMode() const;
+
+		// Invert channel polarity in disabled state
+		constexpr ChannelMode
+		setDisabledPolarity(DisabledPolarity polarity);
+
+		constexpr DisabledPolarity
+		disabledPolarity() const;
+
+		// Configure trigger output source to Timer / Counter peripheral
+		constexpr ChannelMode
+		setTcTriggerSource(TcTriggerSource source);
+
+		constexpr TcTriggerSource
+		tcTriggerSource() const;
+
+		/// Enable or disable dead-time generatiom
+		constexpr ChannelMode
+		setDeadTimeGeneration(DeadTimeGeneration deadTime);
+
+		constexpr DeadTimeGeneration
+		deadTimeGeneration() const;
+
+		/// Set inverted output polarity (after dead-time generator, no effect on override mode)
+		constexpr ChannelMode
+		setOutputPolarity(OutputPolarity polarity);
+
+		constexpr OutputPolarity
+		outputPolarity() const;
+
+		/// Set push-pull mode
+		/// Special mode for transformer-based power converters to avoid core DC saturation.
+		constexpr ChannelMode
+		setPushPullMode(PushPullMode mode);
+
+		constexpr PushPullMode
+		pushPullMode() const;
+
+		/// raw PWM_CMRx register value
+		uint32_t value{};
+	};
```

Comment:
Why new semantics? Isn't this already implemented via the register abstraction?

Reply:
Good point. I have started with an API that would not have worked with register, but ended up with something that does. I'll try to change it.